near-sdk/src/json_types/account.rs

```diff
@@ -33,35 +43,47 @@ impl AsRef<AccountId> for ValidAccountId {
     }
 }
 
-impl<'de> serde::Deserialize<'de> for ValidAccountId {
-    fn deserialize<D>(deserializer: D) -> Result<Self, <D as serde::Deserializer<'de>>::Error>
+impl<'de> Deserialize<'de> for ValidAccountId {
+    fn deserialize<D>(deserializer: D) -> Result<Self, <D as de::Deserializer<'de>>::Error>
     where
-        D: serde::Deserializer<'de>,
+        D: de::Deserializer<'de>,
     {
-        let s = <String as serde::Deserialize>::deserialize(deserializer)?;
-        s.try_into()
-            .map_err(|err: Box<dyn std::error::Error>| serde::de::Error::custom(err.to_string()))
+        let s: String = Deserialize::deserialize(deserializer)?;
+        Self::try_from(s).map_err(de::Error::custom)
     }
 }
 
 impl TryFrom<&str> for ValidAccountId {
     type Error = Box<dyn std::error::Error>;
 
     fn try_from(value: &str) -> Result<Self, Self::Error> {
-        Self::try_from(value.to_string())
+        Self::from_str(value)
+    }
+}
+
+fn validate_account_id(id: &str) -> Result<(), Box<dyn std::error::Error>> {
+    if is_valid_account_id(id.as_bytes()) {
+        Ok(())
+    } else {
+        Err(INVALID_ACCOUNT_ID_MSG.into())
     }
 }
 
 impl TryFrom<String> for ValidAccountId {
     type Error = Box<dyn std::error::Error>;
 
     fn try_from(value: String) -> Result<Self, Self::Error> {
-        let res = Self(value);
-        if res.is_valid() {
-            Ok(res)
-        } else {
-            Err("The account ID is invalid".into())
-        }
+        validate_account_id(value.as_ref())?;
```

Comment:
```suggestion
        validate_account_id(value.as_str())?;
```

To keep this more specific. In general, `as_ref` is a comaptability hazard when used on concrete types, as new impl of AsRef might be added in the future and can regress type inference. That is, use `AsRef` when you have `T: AsRef`, otherwise prefer concrete functions. 

To clarify, this is just a rule of thumb, ther's no specific problems in this case.

Reply:
That's a really good point, I'll do that now!